From 09e30bc99012dd2179706817ecc7468a7c515ba9 Mon Sep 17 00:00:00 2001
From: mayi <3363523343@qq.com>
Date: Thu, 22 Oct 2026 09:52:58 +0800
Subject: [PATCH] fix(osdrv/wifi): avoid ATBM6461 RCU unload deadlock

Problem: CVITEK teardown can wait on the vendor completion before it has been initialized.
Fix: use synchronize_rcu()/rcu_barrier() on this platform.
Prerequisite: driver feature 0009 only; no other optional fix is required.
---
 extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c b/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c
index 4d695aa530296d249a5d9111ad51dcefd161c182..dd011374c29c9481a7fd0ab57fd8d5dccb46a155 100644
--- a/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c
+++ b/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c
@@ -510,2 +510,5 @@ static void atbm_asynchronize_rcu(void)
 {
+#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+	synchronize_rcu();
+#else
 	static int init_cnt = 0;
@@ -518,2 +521,3 @@ static void atbm_asynchronize_rcu(void)
 	init_cnt++;
+#endif
 }
@@ -1627,2 +1631,5 @@ void  apollo_sdio_module_exit(void)
 	atbm_ioctl_free();
+#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+	rcu_barrier();
+#else
 	/* wait rcu sync complete */
@@ -1630,2 +1637,3 @@ void  apollo_sdio_module_exit(void)
 	destroy_rcu_head_on_stack(&atbm_rcu.head);
+#endif
 	#endif
